consensus/istanbul/backend/announce.md

```diff
@@ -0,0 +1,140 @@
+
+- `dest_address`: the destination validator address this message is querying to.
+- `encrypted_enode_url`: the origin's `eNodeURL` encrypted with the destination public key.
+- `version`: the current announce version for the origin's `eNodeURL`.
+- `timestamp`: a message generation timestamp, used to bypass hash caches for messages.
+
+Max amount of encrypted node urls is 2 * size(set of [NearlyElectedValidator]).
+
+### enodeCertificateMsg (0x17)
+
+`[enode_url: B, version: P]`
+
+- `enode_url`: the origin's plaintext `eNodeURL`.
+- `version`: the current announce version for the origin's `eNodeURL`.
+
+### versionCertificatesMsg (0x16)
+
+`[[version: P, signature: B]]`
+
+- `version`: the current highest known announce version for a validator (deduced from the signature), according to the peer that gossipped this message.
+- `signature`: the signature for the version payload string (`'versionCertificate'|version`) by the emmitter of the certificate. Note that the address and public key can be deduced from this `signature`.
+
+## Nearly Elected Validator (NEV)
+
+A validator is a Nearly Elected Validator (NEV) if and only if:
+
+- It is a current elected validator in this epoch, or
+- It is a validator in the result set from calling `ElectNValidatorSigners` with `additionalAboveMaxElectable` taking the value of 10.
+
+## Spec by node type
+
+Note that validators are full nodes too, therefore they should follow the same rules as full nodes. This is specially important to not give away the status of the node as a validator.
+
+### Full node (non validator) Spec
+
+#### Peer handshake
+
+During an inbound peer connection, the remote peer can send an [enodeCertificateMsg] to identify itself as a validator. This allows for preferential treatment for the p2p connection.
+
+#### Peer registered
+
+When a peer is registered, all highest known version certificates should be sent to the registered peer in a [versionCertificatesMsg].
+
+#### Handling [queryEnodeMsg]
+
+Messages received of this type should be only processed once.
+Should be regossipped as is, unless another message from the same validator origin has been regossipped in the past 5 minutes.
+
+#### Handling [enodeCertificateMsg]
+
+[enodeCertificateMsg] should be ignored by non validators.
+
+#### Handling [versionCertificatesMsg]
+
+Messages received of this type should be only processed once.
+When received, regossips a new [versionCertificatesMsg] with certificates that haven't been gossiped (identified by validator address only) in the past 5 minutes
+
+#### Message spawning
+
+Every 5 minutes it should regossip all the highest known version certificates for all the [NearlyElectedValidator]
+
+### Validator Spec
+
+#### Replying [queryEnodeMsg]
+
+Should be replied with an [enodeCertificateMsg] if the origin is a validator and this validator is the destination, and if an already existing connection to the origin validator exists.
```

Comment:
I think we could leave out the bit regarding having an already existing connection. Either way the response should be respond with an enode certificate, whether the connection exists already or whether a new connection needs to be established.

Reply:
As of today, if the connection is not already there, the message is not replied with the enodeCertificateMsg. I assume that the next time the origin validator sends a query, they'll likely be connected, and then the reply will happen. 

What you make me think with this, is that, if the spec does not force to to attempt to connect when receiving a queryEnodeMsg, then this reply may very well never happen, since it's a direct message. So, should the protocol be changed, and enodeCertificateMsg be made into a message that can be safely regossiped (like the queryEnodeIs) using encryption, or should we add to the spec the obligation to 'attempt' to connect when a queryEnodeMsg is received?

Reply:
I think the connection is explicitly made:

If you look here, we can see that if validating the node "multicasts" their enode certificate back to the sender, and that will do nothing if they are not connected, but then at the end of the function it calls `UpsertVersionAndEnode`.

https://github.com/celo-org/celo-blockchain/blob/731fef6f6d390bb3876fe5f5e43d0a0d8d9af6a1/consensus/istanbul/backend/announce.go#L741-L780

At the end of `UpsertVersionAndEnode` `AddValidatorPeer` is called. `AddValidatorPeer` adds a peer to the p2p server with validator purpose, and this will cause the p2p server to try and connect to that peer and when it does because the peer is marked with validator purpose the enode certificate will be sent as part of the handshake.

https://github.com/celo-org/celo-blockchain/blob/731fef6f6d390bb3876fe5f5e43d0a0d8d9af6a1/consensus/istanbul/backend/internal/enodes/val_enode_db.go#L326

Does that look correct to you?

Reply:
Yes, right now when this happens the node attempts to connect to the one that sends the query. But I was talking about the spec, not the implementation. Should the spec explicitly say that a connection should be attempted when receiving the query? If we don't do that, we fall into the problem I've mentioned.

Reply:
I don't think the protocol spec should dictate whether a connection is made or not because that then ties it into a specific implementation. What the spec should make clear is that a message is sent, irrespective of any particular state that the node may be in or the mechanism by which that message is sent.